test/proxy/BeaconProxy.test.js

```diff
@@ -0,0 +1,493 @@
+
+        it('reverts', async function () {
+          const beacon = await UpgradeableBeacon.new(this.implementationV0.address, { from: proxyCreator });
+          await expectRevert(
+            BeaconProxy.new(beacon.address, this.initializeData, { from: proxyCreator, value }),
+            'BeaconProxy: function call failed',
+          );
+        });
+      });
+    });
+
+    describe('payable', function () {
+      const expectedInitializedValue = '42';
+
+      beforeEach(function () {
+        this.initializeData = this.implementationV0.contract.methods
+          .initializePayableWithValue(expectedInitializedValue)
+          .encodeABI();
+      });
+
+      describe('when not sending balance', function () {
+        beforeEach('creating proxy', async function () {
+          const beacon = await UpgradeableBeacon.new(this.implementationV0.address, { from: proxyCreator });
+          this.proxy = (
+            await BeaconProxy.new(beacon.address, this.initializeData, { from: proxyCreator })
+          ).address;
+        });
+
+        assertProxyInitialization({
+          value: expectedInitializedValue,
+          balance: '0',
+        });
+      });
+
+      describe('when sending some balance', function () {
+        const value = 10e5.toString();
+
+        beforeEach('creating proxy', async function () {
+          const beacon = await UpgradeableBeacon.new(this.implementationV0.address, { from: proxyCreator });
+          this.proxy = (
+            await BeaconProxy.new(beacon.address, this.initializeData, {
+              from: proxyCreator,
+              value,
+            })
+          ).address;
+        });
+
+        assertProxyInitialization({
+          value: expectedInitializedValue,
+          balance: value,
+        });
+      });
+    });
+  });
+
+  describe('upgrades', function () {
+    beforeEach(async function () {
+      const initializeData = Buffer.from('');
+      const beacon = await UpgradeableBeacon.new(this.implementationV0.address, { from: proxyCreator });
+      this.proxy = await BeaconProxy.new(beacon.address, initializeData, { from: proxyCreator });
+    });
+
+    describe('implementation', function () {
+      it('returns the current implementation address', async function () {
+        const beacon = new UpgradeableBeacon(await getBeacon(this.proxy.address));
+        const implementation = await beacon.implementation();
+        expect(implementation).to.be.equal(this.implementationV0.address);
+      });
+
+      it('delegates to the implementation', async function () {
+        const dummy = await new DummyImplementation(this.proxy.address);
+        const value = await dummy.get();
+
+        expect(value).to.equal(true);
+      });
+    });
+
+    describe('upgrade', function () {
+      describe('when the sender is the admin', function () {
+        const from = proxyCreator;
+
+        describe('when the given implementation is different from the current one', function () {
```

Comment:
What if it is the same?

Reply:
Well, it will be upgraded too. :slightly_smiling_face: I don't really see the need to test these two scenarios separately. What do you think?

Reply:
Neither do I, but then it doesn't make sense to have this `describe` block if we are not intending to test this as a separate path.